tokio-threadpool/src/notifier.rs

```diff
@@ -19,37 +19,45 @@ impl Notify for Notifier {
     fn notify(&self, id: usize) {
         trace!("Notifier::notify; id=0x{:x}", id);
 
-        let id = id as usize;
-        let task = unsafe { Task::from_notify_id_ref(&id) };
-
-        if !task.schedule() {
-            trace!("    -> task already scheduled");
-            // task is already scheduled, there is nothing more to do
-            return;
-        }
-
-        // TODO: Check if the pool is still running
-
-        // Bump the ref count
-        let task = task.clone();
-
-        if let Some(inner) = self.inner.upgrade() {
-            let _ = inner.submit(task, &inner);
+        unsafe {
+            let ptr = id as *const Task;
+            let task = Arc::from_raw(ptr);
+
+            if task.schedule() {
+                // TODO: Check if the pool is still running
+                //
+                // Bump the ref count
+                let task = task.clone();
+
+                if let Some(inner) = self.inner.upgrade() {
+                    let _ = inner.submit(task, &inner);
+                }
+            }
+
+            // We did not actually take ownership of the `Arc` in this function.
+            mem::forget(task);
         }
     }
 
     fn clone_id(&self, id: usize) -> usize {
         unsafe {
-            let handle = Task::from_notify_id_ref(&id);
-            mem::forget(handle.clone());
+            let ptr = id as *const Task;
+
+            let t1 = Arc::from_raw(ptr);
+            let t2 = t1.clone();
+
+            // Forget both handles as we don't want any ref decs
+            mem::forget(t1);
```

Comment:
please make unsafe code as smaller as possible. mem::forget is a safe function and may be moved out of unsafe block